From a65159b800232f23e9a895f664cae5a7c11d5000 Mon Sep 17 00:00:00 2001
From: Zhang Jiejing <jiejing.zhang@freescale.com>
Date: Wed, 5 Dec 2012 15:13:13 +0800
Subject: [PATCH 06/34] ENGR00235835 add kernel modules build rules.

Add kernel modules build and install rules,
let BoardConfig to install the kernel modules on thier config.

usage:
add the TARGET_KERNEL_MODULES var in your board config:
eg:
(replace '\' to \)

TARGET_KERNEL_MODULES := '\'
    kernel_imx/net/wireless/cfg80211.ko:system/lib/modules/cfg80211.ko

Rules are like the product files.

Signed-off-by: Zhang Jiejing <jiejing.zhang@freescale.com>
(cherry picked from commit 505c9302d54eb9472ce004117b2410c46548f076)
---
 core/Makefile | 29 +++++++++++++++++++++++++++++
 core/main.mk  |  1 +
 2 files changed, 30 insertions(+)

diff --git a/core/Makefile b/core/Makefile
index e661603..e802e4b 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -982,6 +982,35 @@ KERNEL_UIMAGE    := kernel_imx/arch/arm/boot/uImage
 KERNEL_OUT := $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ
 KERNEL_HEADERS_INSTALL := $(KERNEL_OUT)/usr
 
+.PHONY: kernelmodules
+
+kernel_modules_src_list :=
+kernel_modules_dist_list :=
+
+$(foreach cf,$(TARGET_KERNEL_MODULES),							\
+$(eval _src := $(call word-colon,1,$(cf)))						\
+$(eval _dest := $(call word-colon,2,$(cf)))						\
+$(eval kernel_modules_src_list += $(_src))						\
+$(eval kernel_modules_dest_list += $(call append-path,$(PRODUCT_OUT),$(_dest))))
+
+ifneq ($(strip $(kernel_modules_src_list)),)
+
+# kernel modules only depends on kernel config. don't needs kernel have built.
+kernelmodules: $(KERNEL_CONFIGURE) | $(ACP)
+	$(MAKE) -C kernel_imx modules $(KERNEL_ENV)
+	$(hide) $(foreach cf, $(TARGET_KERNEL_MODULES),				\
+	$(eval _src := $(call word-colon,1,$(cf)))				\
+	$(eval _dest := $(call word-colon,2,$(cf)))				\
+	$(eval _full_dest := $(call append-path,$(PRODUCT_OUT),$(_dest)))	\
+	$(eval _dest_dir := $(call dir, $(_full_dest))) 			\
+	mkdir -p $(_dest_dir); 							\
+	$(ACP) $(_src) $(_full_dest);						\
+	)
+	@echo "install kernel modules: $(kernel_modules_src_list)"
+else
+kernelmodules:
+endif
+
 .PHONY: $(TARGET_PREBUILT_KERNEL)
 kernelimage: $(TARGET_PREBUILT_KERNEL)
 
diff --git a/core/main.mk b/core/main.mk
index 8c82f21..c5d330c 100644
--- a/core/main.mk
+++ b/core/main.mk
@@ -876,6 +876,7 @@ endif
 # Build files and then package it into the rom formats
 .PHONY: droidcore
 droidcore: kernelimage \
+	kernelmodules \
 	bootloader \
 	files \
 	systemimage \
-- 
1.8.0

